src/units/unit.cpp

```diff
@@ -406,7 +409,34 @@ unit::unit(unit_ctor_t)
 	, upkeep_(upkeep_full{})
 	, changed_attributes_(0)
 	, invisibility_cache_()
+	, has_ability_distant_(utils::nullopt)
+	, has_ability_distant_image_(utils::nullopt)
 {
+	affect_distant_.clear();
+}
+
+void unit::set_has_ability_distant()
```

Comment:
This is called:
* When the unit is created
* When the unit advances
* When modifications are applied to the unit, which seems to mean `[object]` and for `[modify_unit]`?

Reply:
yes, and in addition the unit creation also uses the advancement function, which generates a double edition, but it remains more profitable to optimize the checking in get_ability_bool()etc, and for modify_unit, the first function call is essential because the third is not called (probably in cases where [effect] is not in a tag [object/advancemnt...)